File: core/InsertDeleteGetrandomO1/InsertDeleteGetrandomO1.py
from random import choice

class RandomizedSet:

    def __init__(self):
        self.nums = []
        self.index = {}

    def insert(self, val: int) -> bool:
        if val in self.index:
            return False
        self.index[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.index:
            return False
        self.nums[self.index[val]] = self.nums[-1]
        self.index[self.nums[-1]] = self.index[val]
        del self.index[val]
        self.nums.pop()
        return True

    def getRandom(self) -> int:
        return choice(self.nums)

File: core/InsertDeleteGetrandomO1/InsertDeleteGetrandomO1.java
package com.dbc;

import java.util.*;

public class InsertDeleteGetrandomO1 {
    private final List<Integer> nums = new ArrayList<>();
    private final Map<Integer, Integer> index = new HashMap<>();
    Random random = new Random();
    public InsertDeleteGetrandomO1() {
    }

    public boolean insert(int val) {
        if (this.index.containsKey(val)) return false;
        this.index.put(val, this.nums.size());
        this.nums.add(val);
        return true;
    }

    public boolean remove(int val) {
        if (!this.index.containsKey(val)) return false;
        this.nums.set(this.index.get(val), this.nums.get(this.nums.size() - 1));
        this.index.put(this.nums.get(this.nums.size() - 1), this.index.get(val));
        this.nums.remove(this.nums.size() - 1);
        this.index.remove(val);
        return true;
    }

    public int getRandom() {
        return this.nums.get(random.nextInt(this.nums.size()));
    }
}
